ugs-core/src/com/willwinder/universalgcodesender/listeners/ControllerStatusBuilder.java

```diff
@@ -107,12 +134,84 @@ public ControllerStatusBuilder setPins(ControllerStatus.EnabledPins pins) {
         return this;
     }
 
+    public static ControllerState getControllerStateFromStateString(String stateString) {
+        return Optional.ofNullable(stateString)
+                       .map(ControllerStatusBuilder::getControllerState)
+                       .orElse(ControllerState.UNKNOWN);
+    }
+
+    private static ControllerState getControllerState(String s) {
+        switch (s.toLowerCase()) {
+            case "jog":
+                return ControllerState.JOG;
+            case "run":
+                return ControllerState.RUN;
+            case "hold":
+                return ControllerState.HOLD;
+            case "door":
+                return ControllerState.DOOR;
+            case "home":
+                return ControllerState.HOME;
+            case "idle":
+                return ControllerState.IDLE;
+            case "alarm":
+                return ControllerState.ALARM;
+            case "check":
+                return ControllerState.CHECK;
+            case "sleep":
+                return ControllerState.SLEEP;
+            default:
+                return ControllerState.UNKNOWN;
+        }
+    }
+
+
     public ControllerStatus build() {
-        return new ControllerStatus(stateString, state, machineCoord, workCoord, feedSpeed, feedSpeedUnits, spindleSpeed, overrides, workCoordinateOffset, pins, states);
+        if (workCoordinateOffset == null) {
+            workCoordinateOffset = Optional.ofNullable(lastStatus)
+                          .map(ControllerStatus::getWorkCoordinateOffset)
+                          .orElseGet(() -> new Position(0, 0, 0, reportingUnits));
+        }
+
+        // Calculate missing coordinate with WCO
+        if (workCoord == null && machineCoord != null) {
+            setWorkCoord(machineCoord.sub(workCoordinateOffset));
+        }
+        if (machineCoord == null && workCoord != null) {
+            setMachineCoord(workCoord.add(workCoordinateOffset));
+        }
+
+        if (lastStatus != null) {
+            feedSpeed = lastStatus.getFeedSpeed();
+            spindleSpeed = lastStatus.getSpindleSpeed();
+        }
+
+        if (overrides != null) {
+            // If this is an override report and the 'Pn:' field wasn't sent
+            // set all pins to a disabled state.
+            if (pins == null) {
+                pins = new ControllerStatus.EnabledPins("");
+            }
+            // Likewise for accessory accessoryStates.
+            if (accessoryStates == null) {
+                accessoryStates = new ControllerStatus.AccessoryStates("");
+            }
+        }
+        else {
+            if (lastStatus != null ) {
+                overrides = lastStatus.getOverrides();
+                pins = lastStatus.getEnabledPins();
+                accessoryStates = lastStatus.getAccessoryStates();
+            }
+        }
+
+        if (!stateOverWritten)
+            state = getControllerStateFromStateString(stateString);
```

Comment:
This will become a problem if I set the state and it gets overwritten by the state string.

Reply:
I think my variable naming isn't the best, this is to prevent your state to be overwritten with state String.